src/main/java/bio/terra/service/resourcemanagement/OneProjectPerResourceSelector.java

```diff
@@ -0,0 +1,107 @@
+import org.springframework.context.annotation.Profile;
+import org.springframework.stereotype.Component;
+import com.google.common.hash.Hashing;
+
+import java.util.UUID;
+
+@Component
+@Profile({"terra", "google"})
+public class OneProjectPerResourceSelector implements DataLocationSelector {
+    private final GoogleResourceConfiguration resourceConfiguration;
+    private final DatasetBucketDao datasetBucketDao;
+    private static final String GS_PROJECT_PATTERN = "[a-z0-9\\-]{6,30}";
+    private static final String GS_BUCKET_PATTERN = "[a-z0-9\\-\\.\\_]{3,63}";
+    /**
+     * *Borrowed from terra-resource-buffer*
+     * The size of project when generating random characters. Choose size as 8 based on AoU's
+     * historical experience, increase if 8 is not enough for a pool's naming.
+     */
+    static final int RANDOM_ID_SIZE = 8;
+
+    @Autowired
+    public OneProjectPerResourceSelector(GoogleResourceConfiguration resourceConfiguration,
+                                         DatasetBucketDao datasetBucketDao) {
+        this.resourceConfiguration = resourceConfiguration;
+        this.datasetBucketDao = datasetBucketDao;
+    }
+
+    @Override
+        public String projectIdForDataset() throws GoogleResourceNamingException {
+        return getNewProjectId();
+    }
+
+    @Override
+    public String projectIdForSnapshot() throws GoogleResourceNamingException {
+        return getNewProjectId();
+    }
+
+    @Override
+    public String projectIdForFile(
+        Dataset dataset, String sourceDatasetGoogleProjectId, BillingProfileModel billingProfile)
+        throws GoogleResourceException, GoogleResourceNamingException {
+        // Case 1
+        // Condition: Requested billing profile matches source dataset's billing profile
+        // Action: Re-use dataset's project
+        UUID sourceDatasetBillingProfileId = dataset.getProjectResource().getProfileId();
+        UUID requestedBillingProfileId = UUID.fromString(billingProfile.getId());
+        if (sourceDatasetBillingProfileId.equals(requestedBillingProfileId)) {
+            return sourceDatasetGoogleProjectId;
+        }
+
+        // Case 2
+        // Condition: Ingest Billing profile != source dataset billing profile && project *already exists*
+        // Action: Re-use bucket's project
+        String bucketGoogleProjectId = datasetBucketDao.getProjectResourceForBucket(dataset.getId(),
+            UUID.fromString(billingProfile.getId()));
+        if (bucketGoogleProjectId != null) {
+            return bucketGoogleProjectId;
+        }
+
+        //Case 3 -
+        // Condition: Ingest Billing profile != source dataset billing profile && project does NOT exist
+        // Action: Create new project
+        return getNewProjectId();
+    }
+
+    @Override
+    public String bucketForFile(String projectId)
+        throws GoogleResourceNamingException {
+        String bucketName = projectId + "-bucket";
```

Comment:
Do we need the `-bucket` suffix here? I thought we were going to use a randomly generated uuid for bucket names: https://github.com/DataBiosphere/jade-data-repo/pull/914/files#diff-eeabaf8988178892d8566d29c291a929bb53298206db69356f948f1db8e93d40L37